### PR TITLE
Improve breadcrumbs implementation [SCI-8087]

### PR DESCRIPTION
**Pull request by sboursen-scinote**
Jira ticket: [SCI-8087](https://scinote.atlassian.net/browse/SCI-8087)

### What was done
- [x] Implemented the new general breadcrumbs in all required views.

[SCI-8087]: https://scinote.atlassian.net/browse/SCI-8087?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/controllers/projects_controller.rb [hunk 1]
```diff
@@ -477,4 +478,33 @@ def log_activity(type_of, project = nil, message_items = {})
...
+    project = @project
+    breadcrumbs_items = []
+    folders = helpers.tree_ordered_parent_folders(current_folder)
+    breadcrumbs_items.push({
+                 label: t('projects.index.breadcrumbs_root'),
+                 url: projects_path,
+                 class: 'project-folder-link'
+               })
+
+    folders&.each do |project_folder|
```
Comment (aignatov-bio) [inline app/controllers/projects_controller.rb:492]: Maybe for PET you can make some concern because this part same for all PET levels

**Review by aignatov-bio — CHANGES_REQUESTED** [review]

File: app/assets/stylesheets/navigation/breadcrumbs.scss [hunk 2]
```diff
@@ -0,0 +1,42 @@
+.sci--layout-navigation-breadcrumbs {
+  --max-breadcrumbs-link-width: 180px;
+  align-items: center;
+  column-gap: 14px;
```
Comment (aignatov-bio) [inline app/assets/stylesheets/navigation/breadcrumbs.scss:4]: use em

File: app/assets/stylesheets/navigation/breadcrumbs.scss [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  display: flex;
+  flex-wrap: wrap;
+  padding: 14px 24px 8px;
```
Comment (aignatov-bio) [inline app/assets/stylesheets/navigation/breadcrumbs.scss:7]: use em

File: app/assets/stylesheets/navigation/breadcrumbs.scss [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    text-overflow: ellipsis;
+    white-space: nowrap;
+
+    &:last-child {
+      color: var(--sn-grey);
+    }
+  }
+
+  .delimiter {
+    padding-bottom: 4px;
```
Comment (aignatov-bio) [inline app/assets/stylesheets/navigation/breadcrumbs.scss:23]: Why you need it?

File: app/controllers/label_templates_controller.rb [hunk 5]
```diff
@@ -186,4 +187,23 @@ def log_activity(type_of, label_template = @label_template, message_items: {})
...
+
+    breadcrumbs_items.push({
+                             label: t('breadcrumbs.labels'),
+                             url: label_templates_path
+                           })
+
+    @breadcrumbs_items = breadcrumbs_items
+
+    if label
+      breadcrumbs_items.push({
```
Comment (aignatov-bio) [inline app/controllers/label_templates_controller.rb:203]: why you need 2 variable local and instance?

File: app/controllers/protocols_controller.rb [hunk 6]
```diff
@@ -1125,4 +1126,24 @@ def protocol_annotation_notification(old_text)
...
+
+    breadcrumbs_items.push({
+                             label: t('breadcrumbs.protocols'),
+                             url: protocols_path
+                           })
+
+    @breadcrumbs_items = breadcrumbs_items
+
+    if protocol
+      breadcrumbs_items.push({
```
Comment (aignatov-bio) [inline app/controllers/protocols_controller.rb:1142]: Same here

File: app/controllers/repositories_controller.rb [hunk 7]
```diff
@@ -525,4 +526,24 @@ def log_activity(type_of, message_items = {})
...
+    breadcrumbs_items = []
+
+    breadcrumbs_items.push({
+                             label: t('breadcrumbs.inventories'),
+                             url: repositories_path
+                           })
+
+    @breadcrumbs_items = breadcrumbs_items
+
+    if repository
```
Comment (aignatov-bio) [inline app/controllers/repositories_controller.rb:541]: Same here, and why you cant use here instance variable

File: app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -57,8 +57,7 @@
     <span style="display: none;" data-hook="body-js"></span>
     <span style="display: none;" data-hook="application-body-html"></span>
 
-
-    <div class="sci--layout" data-breadcrumbs-collapsed="true" data-navigator-collapsed="<%= !content_for?(:sidebar_title) %>">
+    <div class="sci--layout" data-breadcrumbs-collapsed="<%= !@breadcrumbs_items&.length %>" data-navigator-collapsed="<%= !content_for?(:sidebar_title) %>">
```
Comment (aignatov-bio) [inline app/views/layouts/application.html.erb:60]: .present?

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+  <% if shortened %>
+    <% first_breadcrumb_item = @breadcrumbs_items.shift
+      last_breadcrumb_items = @breadcrumbs_items.pop(2) %>
+    <%= link_to(first_breadcrumb_item[:label], first_breadcrumb_item[:url],
+                  class: "breadcrumbs-link #{first_breadcrumb_item[:class]}",
+                  title: first_breadcrumb_item[:label],
+                  data: first_breadcrumb_item[:data]) %>
+    <span class="delimiter">
+      <img src="<%= asset_path "icon_small/navigate_next.svg"%>"
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:13]: use image_tag

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                  title: item[:label],
+                  data: item[:data]) %>
+      <span class="delimiter">
+        <img src="<%= asset_path "icon_small/navigate_next.svg"%>"
+             alt="navigate next"
+             class="navigate_next">
+      </span>
+
+    <% end %>
+    <span class="breadcrumbs-link <%= last_item[:class] %>" title="<%= last_item[:label] %>">
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:64]: i think we don't have anymore custom `class`

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+      <%= last_item[:label] %>
+    </span>
+  <% else %>
+    <% last_item = @breadcrumbs_items.pop %>
+    <% @breadcrumbs_items.each do |item| %>
+      <%= link_to(item[:label], item[:url],
+                  class: "breadcrumbs-link #{item[:class]}",
+                  title: item[:label],
+                  data: item[:data]) %>
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:56]: Do we have data?

File: app/controllers/concerns/breadcrumbs.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+      project = experiment&.project || @project
+      current_folder = project&.project_folder || @current_folder
+      @breadcrumbs_items = []
+
+      folders = helpers.tree_ordered_parent_folders(current_folder)
+
+      @breadcrumbs_items.push({
+                                label: t('projects.index.breadcrumbs_root'),
+                                url: projects_path,
+                                class: 'project-folder-link'
```
Comment (aignatov-bio) [inline app/controllers/concerns/breadcrumbs.rb:19]: Do we need this class?

**Review by aignatov-bio — CHANGES_REQUESTED** [review]

File: app/assets/stylesheets/navigation/breadcrumbs.scss (hunk 4 above)
Comment (sboursen-scinote) [reply app/assets/stylesheets/navigation/breadcrumbs.scss:23]: Yes. without it, it doesn't look like figma

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    <span class="delimiter">
+      <%= image_tag "icon_small/navigate_next.svg",
+                    alt: "navigate next",
+                    class: "navigate_next" %>
+    </span>
+    <% item = last_breadcrumb_items.first %>
+    <%= link_to(item[:label], item[:url],
+                class: "breadcrumbs-link #{item[:class]}",
+                title: item[:label],
+                data: item[:data]) %>
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:40]: for what data used? ?

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        </ul>
+    </span>
+    <span class="delimiter">
+      <%= image_tag "icon_small/navigate_next.svg",
+                    alt: "navigate next",
+                    class: "navigate_next" %>
+    </span>
+    <% item = last_breadcrumb_items.first %>
+    <%= link_to(item[:label], item[:url],
+                class: "breadcrumbs-link #{item[:class]}",
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:38]: Why class?

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    </span>
+    <% last_item = last_breadcrumb_items.last %>
+    <span class="breadcrumbs-link <%= last_item[:class] %>" title="<%= last_item[:label] %>">
+      <%= last_item[:label] %>
+    </span>
+  <% else %>
+    <% last_item = @breadcrumbs_items.pop %>
+    <% @breadcrumbs_items.each do |item| %>
+      <%= link_to(item[:label], item[:url],
+                  class: "breadcrumbs-link #{item[:class]}",
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:54]: class ?

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    <span class="breadcrumbs-collapsed-container">
+        <span class="show-breadcrumbs" data-toggle="dropdown" title="<%= t('projects.index.breadcrumbs_collapsed') %>">
+          •••
+          <span class="caret pull-right"></span>
+        </span>
+        <ul class="dropdown-menu breadcrumbs-dropdown" role="menu">
+          <% @breadcrumbs_items.each do |item| %>
+            <li>
+              <%= link_to(item[:label], item[:url],
+                          data: item[:data]) %>
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:26]: data?

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+  <% @breadcrumbs_items.each { |item| item[:label] = "(A) #{item[:label]}" if item[:archived] }%>
+  <% shortened = @breadcrumbs_items.length > 4 %>
+
+  <% if shortened %>
+    <% first_breadcrumb_item = @breadcrumbs_items.shift
+      last_breadcrumb_items = @breadcrumbs_items.pop(2) %>
+    <%= link_to(first_breadcrumb_item[:label], first_breadcrumb_item[:url],
+                  class: "breadcrumbs-link #{first_breadcrumb_item[:class]}",
+                  title: first_breadcrumb_item[:label],
+                  data: first_breadcrumb_item[:data]) %>
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:11]: data?

File: app/views/shared/navigation/_breadcrumbs.html.erb [hunk 18]
```diff
@@ -0,0 +1,67 @@
+<% if @breadcrumbs_items&.length %>
+  <% @breadcrumbs_items.each { |item| item[:label] = "(A) #{item[:label]}" if item[:archived] }%>
+  <% shortened = @breadcrumbs_items.length > 4 %>
+
+  <% if shortened %>
+    <% first_breadcrumb_item = @breadcrumbs_items.shift
+      last_breadcrumb_items = @breadcrumbs_items.pop(2) %>
+    <%= link_to(first_breadcrumb_item[:label], first_breadcrumb_item[:url],
+                  class: "breadcrumbs-link #{first_breadcrumb_item[:class]}",
```
Comment (aignatov-bio) [inline app/views/shared/navigation/_breadcrumbs.html.erb:9]: class?

**Review by aignatov-bio — CHANGES_REQUESTED** [review]

**Review by aignatov-bio — APPROVED** [review]

**Pull request merged**